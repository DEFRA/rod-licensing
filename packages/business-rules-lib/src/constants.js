/**
 * The maximum days in advance of the purchase date that the licence may start
 * @type {number}
 */
export const ADVANCED_PURCHASE_MAX_DAYS = 30

/**
 * The  maximum number of permissions in a (multi-buy) transaction
 * @type {number}
 */
export const MAX_PERMISSIONS_PER_TRANSACTION = 50

/**
 * Allowed status values for the payment journal
 */
export const PAYMENT_JOURNAL_STATUS_CODES = {
  InProgress: 'In Progress',
  Cancelled: 'Cancelled',
  Failed: 'Failed',
  Expired: 'Expired',
  Completed: 'Completed'
}

/**
 * Payment status returned by the GOV.UK pay API
 */
export const PAYMENT_STATUS = {
  Success: 'success',
  Failure: 'failed'
}

/**
 * Failure states returned by the GOV.UK pay API
 * https://docs.payments.service.gov.uk/api_reference/#errors-caused-by-payment-statuses
 */
export const GOVUK_PAY_ERROR_STATUS_CODES = {
  REJECTED: 'P0010',
  EXPIRED: 'P0020',
  USER_CANCELLED: 'P0030',
  ERROR: 'P0050',
  NOT_FOUND: 'P0200'
}

/**
 * Constants used in finalization
 * @type {{telesales: string, govPay: string}}
 */
export const TRANSACTION_SOURCE = { govPay: 'Gov Pay', telesales: 'Telesales' }
export const PAYMENT_TYPE = { debit: 'Debit card' }

/**
 * Data sources associated with POCL transactions
 */
export const POCL_DATA_SOURCE = 'Post Office Sales'
export const DDE_DATA_SOURCE = 'DDE File'
export const POCL_TRANSACTION_SOURCES = [POCL_DATA_SOURCE, DDE_DATA_SOURCE]

/**
 * Describes the lifetime of the renewal link.
 */
export const RENEW_BEFORE_DAYS = 60 // The number of days before licence expiry the link is active
export const RENEW_AFTER_DAYS = 60 // The number of days after licence expiry the link is active

/**
 * The number of minutes after payment that a licence will start
 * @type {number}
 */
export const START_AFTER_PAYMENT_MINUTES = 30

/**
 * Timezone of the service
 */
export const SERVICE_LOCAL_TIME = 'Europe/London'

/**
 * Date for switching paper fulfilment provider
 */
export const FULFILMENT_SWITCHOVER_DATE = '2024-05-30T23:00:00.000Z'
