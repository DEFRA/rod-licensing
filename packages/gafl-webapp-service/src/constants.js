/**
 * System constants and defaults
 */
export const ADDRESS_LOOKUP_SERVICE = { lang: 'EN', dataset: 'DPA' }
export const ADDRESS_LOOKUP_TIMEOUT_MS_DEFAULT = 10000
export const ATTRIBUTION_REDIRECT_DEFAULT = '/'
export const SESSION_TTL_MS_DEFAULT = 3 * 60 * 60 * 1000
export const PORT_DEFAULT = 3000
export const REDIS_PORT_DEFAULT = 6379
export const PAGE_STATE = { completed: true, error: false }
export const SESSION_COOKIE_NAME_DEFAULT = 'sid'
export const CSRF_TOKEN_COOKIE_NAME_DEFAULT = 'rlsctkn'
export const FEEDBACK_URI_DEFAULT = '#'
export const CHANNEL_DEFAULT = 'websales'
export const SERVICE_PAGE_DEFAULT = 'https://www.gov.uk/fishing-licences/buy-a-fishing-licence'

export const COMPLETION_STATUS = {
  agreed: 'agreed',
  posted: 'posted',
  paymentCreated: 'payment-created',
  paymentCancelled: 'payment-cancelled',
  paymentFailed: 'payment-failed',
  paymentCompleted: 'payment-completed',
  finalised: 'finalised',
  completed: 'completed'
}

export const GOVPAYFAIL = {
  prePaymentRetry: { step: 'pre-payment' },
  postPaymentRetry: { step: 'post-payment' }
}

export const UTM = {
  CAMPAIGN: 'utm_campaign',
  MEDIUM: 'utm_medium',
  SOURCE: 'utm_source',
  TERM: 'utm_term',
  CONTENT: 'utm_content'
}

export const dateFormats = ['YYYY-MM-DD', 'YY-MM-DD', 'YYYY-M-DD', 'YY-M-DD', 'YYYY-MM-D', 'YY-MM-D', 'YYYY-M-D', 'YY-M-D']

export const CommonResults = {
  SUMMARY: 'summary',
  OK: 'ok',
  NO: 'no',
  YES: 'yes'
}

export const allowsPhysicalLicence = {
  YES: 'yes',
  NO: 'no'
}

// If the user has seen the summary page these are set in the status
export const CONTACT_SUMMARY_SEEN = 'contact-summary'
export const LICENCE_SUMMARY_SEEN = 'licence-summary'

// These cookies are used by the load balancer
export const ALB_COOKIE_NAME = 'AWSALBTG'
export const ALBCORS_COOKIE_NAME = 'AWSALBTGCORS'
