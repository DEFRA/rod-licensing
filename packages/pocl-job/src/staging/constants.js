export const MAX_CREATE_TRANSACTION_BATCH_SIZE = 25
export const MAX_FINALISE_TRANSACTION_BATCH_SIZE = 10

export const POST_OFFICE_DATASOURCE = 'Post Office Sales'
export const DIRECT_DEBIT_DATASOURCE = 'DDE File'
export const DIRECT_DEBIT_PAYMENTSOURCE = 'Direct Debit'
export const POSTAL_ORDER_DATASOURCE = 'Postal Order Sales'
export const POSTAL_ORDER_PAYMENTSOURCE = 'Postal Order'

export const FILE_STAGE = {
  Pending: 'Received and Pending',
  Staging: 'Staging Transactions',
  Finalising: 'Finalising Transactions',
  Completed: 'Completed'
}

export const DYNAMICS_IMPORT_STAGE = {
  Pending: 'Received and Pending',
  InProgress: 'In Progress',
  Processed: 'Processed',
  ProcessedWithWarnings: 'Processed with Warnings',
  Failed: 'Failed',
  UnderReview: 'Under Review',
  Closed: 'Closed',
  isAlreadyProcessed: status => status !== DYNAMICS_IMPORT_STAGE.Pending && status !== DYNAMICS_IMPORT_STAGE.InProgress
}

export const RECORD_STAGE = {
  TransactionCreated: 'Transaction Created',
  TransactionCreationFailed: 'Transaction Creation Failed',
  TransactionFinalised: 'Transaction Finalised',
  TransactionFinalisationFailed: 'Transaction Finalisation Failed'
}
