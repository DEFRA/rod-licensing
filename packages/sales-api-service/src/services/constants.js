const CONCESSION = {
  SENIOR: 'Senior',
  JUNIOR: 'Junior',
  DISABLED: 'Disabled'
}

const CONCESSION_PROOF = {
  none: 'No Proof',
  blueBadge: 'Blue Badge',
  NI: 'National Insurance Number'
}

export { CONCESSION, CONCESSION_PROOF }
