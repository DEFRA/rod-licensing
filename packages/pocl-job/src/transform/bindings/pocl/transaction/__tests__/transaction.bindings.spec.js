import { Transaction } from '../transaction.bindings.js'
import { POST_OFFICE_DATASOURCE, DIRECT_DEBIT_DATASOURCE } from '../../../../../staging/constants.js'

jest.mock('@defra-fish/connectors-lib', () => ({
  salesApi: {
    ...Object.keys(jest.requireActual('@defra-fish/connectors-lib').salesApi).reduce((acc, k) => ({ ...acc, [k]: jest.fn() }), {}),
    ...['permits', 'concessions'].reduce(
      (acc, m) => ({ ...acc, [m]: { find: jest.fn(() => ({ id: `${m}-id`, isForFulfilment: false })) } }),
      {}
    )
  }
}))

describe('transaction transforms', () => {
  it('transforms a POCL record - gmt time, no concession, email contact', async () => {
    const result = await Transaction.transform(generateInputJSON())

    expect(result).toMatchSnapshot()
  })

  it('transforms a POCL record - bst time, senior concession, sms contact', async () => {
    const result = await Transaction.transform(
      generateInputJSON({
        DOB: { value: '12/03/1931' },
        SENIOR_ID: { value: 'Uncancelled Passport' },
        NOTIFY_EMAIL_ADDRESS: { value: '' },
        COMMS_EMAIL_ADDRESS: { value: '' },
        NOTIFY_SMS_NUMBER: { value: '07124567890' },
        COMMS_SMS_NUMBER: { value: '07124567891' },
        NOTIFY_EMAIL: { value: 'N' },
        NOTIFY_SMS: { value: 'Y' },
        COMMS_EMAIL: { value: 'N' },
        COMMS_SMS: { value: 'Y' },
        START_DATE: { value: '01/06/2020' },
        SYSTEM_DATE: { value: '01/06/2020' },
        SYSTEM_TIME: { value: '09:00:00' },
        MOPEX: { value: 'UNKNOWN' }
      })
    )

    expect(result).toMatchSnapshot()
  })

  it('transforms a POCL record - disabled concession (both), letter contact', async () => {
    const result = await Transaction.transform(
      generateInputJSON({
        DOB: { value: '12/03/1931' },
        SENIOR_ID: { value: '' },
        DISABLED_ID_1: { value: '12ABCD01234X5678' },
        DISABLED_ID_2: { value: 'QQ123456C' },
        NOTIFY_EMAIL_ADDRESS: { value: '' },
        COMMS_EMAIL_ADDRESS: { value: '' },
        NOTIFY_SMS_NUMBER: { value: '' },
        COMMS_SMS_NUMBER: { value: '' },
        NOTIFY_EMAIL: { value: 'N' },
        NOTIFY_POST: { value: 'Y' },
        COMMS_EMAIL: { value: 'N' },
        COMMS_POST: { value: 'Y' }
      })
    )

    expect(result).toMatchSnapshot()
  })

  it('transforms a POCL record - disabled concession (blue badge only), no contact', async () => {
    const result = await Transaction.transform(
      generateInputJSON({
        DOB: { value: '12/03/1931' },
        SENIOR_ID: { value: '' },
        DISABLED_ID_1: { value: '12ABCD01234X5678' },
        DISABLED_ID_2: { value: '' },
        NOTIFY_EMAIL_ADDRESS: { value: '' },
        COMMS_EMAIL_ADDRESS: { value: '' },
        NOTIFY_SMS_NUMBER: { value: '' },
        COMMS_SMS_NUMBER: { value: '' },
        NOTIFY_EMAIL: { value: 'N' },
        COMMS_EMAIL: { value: 'N' }
      })
    )

    expect(result).toMatchSnapshot()
  })

  it('transforms a POCL record - disabled concession (pip only), multiple contact', async () => {
    const result = await Transaction.transform(
      generateInputJSON({
        DOB: { value: '12/03/1931' },
        SENIOR_ID: { value: '' },
        DISABLED_ID_1: { value: '' },
        DISABLED_ID_2: { value: 'QQ123456C' },
        NOTIFY_EMAIL_ADDRESS: { value: '' },
        COMMS_EMAIL_ADDRESS: { value: 'festerthetester456@email.com' },
        COMMS_SMS_NUMBER: { value: '07124567891' },
        NOTIFY_EMAIL: { value: 'N' },
        NOTIFY_SMS: { value: 'Y' }
      })
    )

    expect(result).toMatchSnapshot()
  })

  it('transforms a DDE record - gmt time, no concession, email contact', async () => {
    const result = await Transaction.transform(
      generateInputJSON({
        DATA_SOURCE: { value: DIRECT_DEBIT_DATASOURCE },
        MOPEX: { value: '6' }
      })
    )

    expect(result).toMatchSnapshot()
  })

  it('ignores invalid datasource', async () => {
    const result = await Transaction.transform(
      generateInputJSON({
        DATA_SOURCE: { value: 'Any old rubbish' }
      })
    )

    expect(result).toEqual(
      expect.objectContaining({
        createTransactionPayload: expect.objectContaining({
          dataSource: POST_OFFICE_DATASOURCE
        }),
        finaliseTransactionPayload: {
          payment: expect.objectContaining({
            source: POST_OFFICE_DATASOURCE
          })
        }
      })
    )
  })
})

const generateInputJSON = (overrides = {}) => ({
  children: {
    LICENSEE_FORENAME: { value: 'Fester' },
    LICENSEE_SURNAME: { value: 'Tester' },
    DOB: { value: '12/03/1981' },
    NOTIFY_EMAIL_ADDRESS: { value: 'festerthetester123@email.com' },
    COMMS_EMAIL_ADDRESS: { value: 'festerthetester456@email.com' },
    NOTIFY_EMAIL: { value: 'Y' },
    NOTIFY_SMS: { value: 'N' },
    NOTIFY_POST: { value: 'N' },
    COMMS_EMAIL: { value: 'Y' },
    COMMS_SMS: { value: 'N' },
    COMMS_POST: { value: 'N' },
    LICENSEE_ADDRESS: {
      children: {
        POBox: { value: 'PO Box 123' },
        Subprem: { value: 'Flat 5' },
        Buildname: { value: 'Angling House' },
        Buildnum: { value: '123' },
        Depthoro: { value: 'Redhill Park' },
        Thoro: { value: 'Redhill Lane' },
        Deplocal: { value: 'Lower Denton' },
        Local: { value: 'Denton' },
        Town: { value: 'Manchester' },
        Postcode: { value: 'AB123CD' }
      }
    },
    ITEM_ID: { value: '123456' },
    START_DATE: { value: '01/01/2020' },
    START_TIME: { value: '00:01' },
    SYSTEM_DATE: { value: '01/01/2020' },
    SYSTEM_TIME: { value: '15:03:17' },
    AMOUNT: { value: '3.75' },
    MOPEX: { value: '1' },
    SERIAL_NO: { value: '559136-2-27950' },
    CHANNEL_ID: { value: '123456' },
    ...overrides
  }
})
