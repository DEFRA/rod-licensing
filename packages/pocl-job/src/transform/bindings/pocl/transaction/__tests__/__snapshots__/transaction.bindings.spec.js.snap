// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transaction transforms transforms a DDE record - gmt time, no concession, email contact 1`] = `
Object {
  "createTransactionPayload": Object {
    "dataSource": "DDE File",
    "permissions": Array [
      Object {
        "issueDate": "2020-01-01T15:03:17.000Z",
        "licensee": Object {
          "birthDate": "1981-03-12",
          "country": "GB",
          "email": "festerthetester123@email.com",
          "firstName": "Fester",
          "lastName": "Tester",
          "locality": "Lower Denton, Denton",
          "postalFulfilment": false,
          "postcode": "AB123CD",
          "preferredMethodOfConfirmation": "Email",
          "preferredMethodOfNewsletter": "Email",
          "preferredMethodOfReminder": "Email",
          "premises": "PO Box 123, Flat 5, Angling House, 123",
          "street": "Redhill Park, Redhill Lane",
          "town": "Manchester",
        },
        "permitId": "permits-id",
        "startDate": "2020-01-01T00:01:00.000Z",
      },
    ],
    "serialNumber": "559136-2-27950",
  },
  "finaliseTransactionPayload": Object {
    "payment": Object {
      "amount": 3.75,
      "channelId": "123456",
      "method": "Direct debit",
      "source": "Direct Debit",
      "timestamp": "2020-01-01T15:03:17.000Z",
    },
  },
  "id": "559136-2-27950",
}
`;

exports[`transaction transforms transforms a POCL record - bst time, senior concession, sms contact 1`] = `
Object {
  "createTransactionPayload": Object {
    "dataSource": "Post Office Sales",
    "permissions": Array [
      Object {
        "concessions": Array [
          Object {
            "id": "concessions-id",
            "proof": Object {
              "referenceNumber": "N/A",
              "type": "Uncancelled Passport",
            },
          },
        ],
        "issueDate": "2020-06-01T08:00:00.000Z",
        "licensee": Object {
          "birthDate": "1931-03-12",
          "country": "GB",
          "firstName": "Fester",
          "lastName": "Tester",
          "locality": "Lower Denton, Denton",
          "mobilePhone": "07124567890",
          "postalFulfilment": false,
          "postcode": "AB123CD",
          "preferredMethodOfConfirmation": "Text",
          "preferredMethodOfNewsletter": "Text",
          "preferredMethodOfReminder": "Text",
          "premises": "PO Box 123, Flat 5, Angling House, 123",
          "street": "Redhill Park, Redhill Lane",
          "town": "Manchester",
        },
        "permitId": "permits-id",
        "startDate": "2020-05-31T23:01:00.000Z",
      },
    ],
    "serialNumber": "559136-2-27950",
  },
  "finaliseTransactionPayload": Object {
    "payment": Object {
      "amount": 3.75,
      "channelId": "123456",
      "method": "Other",
      "source": "Post Office Sales",
      "timestamp": "2020-06-01T08:00:00.000Z",
    },
  },
  "id": "559136-2-27950",
}
`;

exports[`transaction transforms transforms a POCL record - disabled concession (blue badge only), no contact 1`] = `
Object {
  "createTransactionPayload": Object {
    "dataSource": "Post Office Sales",
    "permissions": Array [
      Object {
        "concessions": Array [
          Object {
            "id": "concessions-id",
            "proof": Object {
              "referenceNumber": "12ABCD01234X5678",
              "type": "Blue Badge",
            },
          },
        ],
        "issueDate": "2020-01-01T15:03:17.000Z",
        "licensee": Object {
          "birthDate": "1931-03-12",
          "country": "GB",
          "firstName": "Fester",
          "lastName": "Tester",
          "locality": "Lower Denton, Denton",
          "postalFulfilment": false,
          "postcode": "AB123CD",
          "preferredMethodOfConfirmation": "Prefer not to be contacted",
          "preferredMethodOfNewsletter": "Prefer not to be contacted",
          "preferredMethodOfReminder": "Prefer not to be contacted",
          "premises": "PO Box 123, Flat 5, Angling House, 123",
          "street": "Redhill Park, Redhill Lane",
          "town": "Manchester",
        },
        "permitId": "permits-id",
        "startDate": "2020-01-01T00:01:00.000Z",
      },
    ],
    "serialNumber": "559136-2-27950",
  },
  "finaliseTransactionPayload": Object {
    "payment": Object {
      "amount": 3.75,
      "channelId": "123456",
      "method": "Cash",
      "source": "Post Office Sales",
      "timestamp": "2020-01-01T15:03:17.000Z",
    },
  },
  "id": "559136-2-27950",
}
`;

exports[`transaction transforms transforms a POCL record - disabled concession (both), letter contact 1`] = `
Object {
  "createTransactionPayload": Object {
    "dataSource": "Post Office Sales",
    "permissions": Array [
      Object {
        "concessions": Array [
          Object {
            "id": "concessions-id",
            "proof": Object {
              "referenceNumber": "12ABCD01234X5678",
              "type": "Blue Badge",
            },
          },
        ],
        "issueDate": "2020-01-01T15:03:17.000Z",
        "licensee": Object {
          "birthDate": "1931-03-12",
          "country": "GB",
          "firstName": "Fester",
          "lastName": "Tester",
          "locality": "Lower Denton, Denton",
          "postalFulfilment": false,
          "postcode": "AB123CD",
          "preferredMethodOfConfirmation": "Letter",
          "preferredMethodOfNewsletter": "Letter",
          "preferredMethodOfReminder": "Letter",
          "premises": "PO Box 123, Flat 5, Angling House, 123",
          "street": "Redhill Park, Redhill Lane",
          "town": "Manchester",
        },
        "permitId": "permits-id",
        "startDate": "2020-01-01T00:01:00.000Z",
      },
    ],
    "serialNumber": "559136-2-27950",
  },
  "finaliseTransactionPayload": Object {
    "payment": Object {
      "amount": 3.75,
      "channelId": "123456",
      "method": "Cash",
      "source": "Post Office Sales",
      "timestamp": "2020-01-01T15:03:17.000Z",
    },
  },
  "id": "559136-2-27950",
}
`;

exports[`transaction transforms transforms a POCL record - disabled concession (pip only), multiple contact 1`] = `
Object {
  "createTransactionPayload": Object {
    "dataSource": "Post Office Sales",
    "permissions": Array [
      Object {
        "concessions": Array [
          Object {
            "id": "concessions-id",
            "proof": Object {
              "referenceNumber": "QQ123456C",
              "type": "National Insurance Number",
            },
          },
        ],
        "issueDate": "2020-01-01T15:03:17.000Z",
        "licensee": Object {
          "birthDate": "1931-03-12",
          "country": "GB",
          "email": "festerthetester456@email.com",
          "firstName": "Fester",
          "lastName": "Tester",
          "locality": "Lower Denton, Denton",
          "mobilePhone": "07124567891",
          "postalFulfilment": false,
          "postcode": "AB123CD",
          "preferredMethodOfConfirmation": "Text",
          "preferredMethodOfNewsletter": "Email",
          "preferredMethodOfReminder": "Text",
          "premises": "PO Box 123, Flat 5, Angling House, 123",
          "street": "Redhill Park, Redhill Lane",
          "town": "Manchester",
        },
        "permitId": "permits-id",
        "startDate": "2020-01-01T00:01:00.000Z",
      },
    ],
    "serialNumber": "559136-2-27950",
  },
  "finaliseTransactionPayload": Object {
    "payment": Object {
      "amount": 3.75,
      "channelId": "123456",
      "method": "Cash",
      "source": "Post Office Sales",
      "timestamp": "2020-01-01T15:03:17.000Z",
    },
  },
  "id": "559136-2-27950",
}
`;

exports[`transaction transforms transforms a POCL record - gmt time, no concession, email contact 1`] = `
Object {
  "createTransactionPayload": Object {
    "dataSource": "Post Office Sales",
    "permissions": Array [
      Object {
        "issueDate": "2020-01-01T15:03:17.000Z",
        "licensee": Object {
          "birthDate": "1981-03-12",
          "country": "GB",
          "email": "festerthetester123@email.com",
          "firstName": "Fester",
          "lastName": "Tester",
          "locality": "Lower Denton, Denton",
          "postalFulfilment": false,
          "postcode": "AB123CD",
          "preferredMethodOfConfirmation": "Email",
          "preferredMethodOfNewsletter": "Email",
          "preferredMethodOfReminder": "Email",
          "premises": "PO Box 123, Flat 5, Angling House, 123",
          "street": "Redhill Park, Redhill Lane",
          "town": "Manchester",
        },
        "permitId": "permits-id",
        "startDate": "2020-01-01T00:01:00.000Z",
      },
    ],
    "serialNumber": "559136-2-27950",
  },
  "finaliseTransactionPayload": Object {
    "payment": Object {
      "amount": 3.75,
      "channelId": "123456",
      "method": "Cash",
      "source": "Post Office Sales",
      "timestamp": "2020-01-01T15:03:17.000Z",
    },
  },
  "id": "559136-2-27950",
}
`;
